### PR TITLE
Refactor class Settings, step 2

### PR DESCRIPTION
**Pull request by ksooo**
Fixes Coverity complaints after https://github.com/kodi-pvr/pvr.hts/commit/f013946b78e5b4770316883edfdf8028140d7d0a @Jalle19 mind taking a look.

BTW: I overlooked the public members in class Settings you've just introduced. I do not like that, tbh and would have bugged you with that if I had seen this before the merge. ;-)

### CODE REVIEW

**Comment by Jalle19** [comment]
Yeah leaving them public was a quick fix, we should change them to use getters/setters like the other new classes.

File: src/tvheadend/Settings.h [hunk 1]
```diff
@@ -55,7 +55,10 @@ namespace tvheadend {
     int         iAutorecMaxDiff;
 
   private:
-    Settings() { }
+    Settings()
+    : iPortHTSP(9981), iPortHTTP(9982), iConnectTimeout(10), iResponseTimeout(5),
```
Comment (Jalle19) [inline src/tvheadend/Settings.h]: There are #defines for these somewhere I think?

Comment (ksooo) [inline src/tvheadend/Settings.h]: Good point.

**Comment by ksooo** [comment]
@Jalle19 now with proper defaults, encapsulated members... Could you please review?

File: src/HTSPConnection.cpp [hunk 2]
```diff
@@ -115,13 +115,13 @@ std::string CHTSPConnection::GetWebURL ( const char *fmt, ... )
   const Settings &settings = Settings::GetInstance();
 
   // Generate the authentication string (user:pass@)
-  std::string auth = settings.strUsername;
-  if (!(auth.empty() || settings.strPassword.empty()))
-    auth += ":" + settings.strPassword;
+  std::string auth = settings.GetUsername();
+  if (!(auth.empty() || settings.GetPassword().empty()))
+    auth += ":" + settings.GetPassword();
```
Comment (Jalle19) [inline src/HTSPConnection.cpp]: Now that `Settings` is an actual class with actual methods I was thinking we could factor away some things like this to separate methods, e.g. `GetAuthenticationString()` in this case. This may be an edge case since it's technically not that related to "settings" so it's up to you if you want to do it.

File: src/client.cpp [hunk 3]
```diff
@@ -142,33 +141,26 @@ ADDON_STATUS ADDON_Create(void* hdl, void* _unused(props))
...
-    settings.iPreTuneCloseDelay = g_iPreTunerCloseDelay;
-  }
-  else
-  {
-    /* When we don't want to use predictive tuning */
-    settings.iTotalTuners = 1;
-    settings.iPreTuneCloseDelay = 0;
-  }
+  settings.SetPreTunerEnabled(g_bPreTunerEnabled);
+  settings.SetTotalTuners(g_bPreTunerEnabled ? g_iTotalTuners : 1);
```
Comment (Jalle19) [inline src/client.cpp]: Similar thing here, IMO `SetTotalTuners` could handle this check internally. It's one of the benefits of setters compared to plain assignment.

File: src/tvheadend/Settings.h [hunk 4]
```diff
@@ -31,6 +31,22 @@ namespace tvheadend {
   class Settings {
   public:
 
+    // Default values.
+    static const std::string DEFAULT_HOST;
+    static const int         DEFAULT_HTTP_PORT           = 9981;
```
Comment (Jalle19) [inline src/tvheadend/Settings.h]: IMO it would make sense to define all these values in the newly introduced `.cpp` file, currently if you're looking at just the `.cpp` file you may come under the impression that those three constants are the only ones available in the class. Up to you though.

File: src/tvheadend/entity/Entity.h [hunk 5]
```diff
@@ -34,7 +34,7 @@ namespace tvheadend
     class Entity
     {
     public:
-      Entity() : m_dirty(false), m_id(0) {};
+      Entity() : m_id(0), m_dirty(false) {};
```
Comment (Jalle19) [inline src/tvheadend/entity/Entity.h]: If you find yourself wondering what to do on a Friday night you could perhaps move this change to a separate commit ;-)

**Comment by Jalle19** [comment]
Apart from the comments, very nice!

File: src/tvheadend/entity/Entity.h (hunk 5 above)
Comment (ksooo) [inline src/tvheadend/entity/Entity.h]: This is already a separate commit. ;-)

Comment (Jalle19) [inline src/tvheadend/entity/Entity.h]: WTF you're totally right

File: src/tvheadend/Settings.h (hunk 4 above)
Comment (ksooo) [inline src/tvheadend/Settings.h]: Makes sense.

**Comment by ksooo** [comment]
@Jalle19 for what do we need these g_blah settings vars? Could we not just get rid of those?

**Comment by Jalle19** [comment]
@ksooo nothing really, we really should get rid of them and the `UPDATE_STR`/`UPDATE_INT` macros at the same time and just handle "did a setting just change" in the new `Settings` class instead.

**Comment by ksooo** [comment]
yep. will do that.

**Comment by Jalle19** [comment]
Thanks, just tack it onto this PR.

**Comment by ksooo** [comment]
@Jalle19 consider this ready now. Mind doing final review?

File: src/tvheadend/Settings.cpp [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  /* Data Transfer */
+  else if (key == "epg_async")
+    return SetBoolSetting(GetTraceDebug(), value);
+  /* Predictive Tuning */
+  else if (key == "pretuner_enabled")
+  {
+    if (GetTotalTuners() > 1 && *(reinterpret_cast<const bool*>(value)))
+      return ADDON_STATUS_OK;
+    else
+     return ADDON_STATUS_NEED_RESTART;
```
Comment (Jalle19) [inline src/tvheadend/Settings.cpp]: Missing one space

File: src/tvheadend/Settings.cpp [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+  /* Auto recordings */
+  SetAutorecApproxTime(ReadIntSetting("autorec_approxtime", DEFAULT_APPROX_TIME));
+  SetAutorecMaxDiff(ReadIntSetting("autorec_maxdiff", DEFAULT_AUTOREC_MAXDIFF));
+}
+
+ADDON_STATUS Settings::SetStringSetting(const std::string &oldValue, const void *newValue)
+{
+  if (oldValue == std::string(reinterpret_cast<const char *>(newValue)))
+  {
+    // value not changed.
```
Comment (Jalle19) [inline src/tvheadend/Settings.cpp]: Minor thing, but if you'd add these comments in the method documentation in the header file then this method and the two below it could be shortened to 4 lines each instead of the current 10.

**Comment by Jalle19** [comment]
Very nice, it's good to finally get rid of the globals and those pesky macros (even though it all worked just fine).

**Comment by Jalle19** [comment]
@ksooo push the button when you're ready, the last comments aren't that important since they're just cosmetics.

**Comment by ksooo** [comment]
Next step: Rip logging stuff out of Tvheadend.h. Currently, Settings.cpp needs to include Tvheadend.h only because of using tvherror.

**Comment by Jalle19** [comment]
@ksooo I actually started work on that last night. I would like to have e.g. `Logger::LogInfo("blabla %s", "foo")`. Ideally I'd like to inject the `XBMC->Log(...)` part through a callback so the logger doesn't explicitly have to include `client.h` to get access to that global (because you can't have multiple `libXBMC_addonHelper` objects in your code.

**Comment by ksooo** [comment]
This is actually literally what I had in mind. So, your turn?

**Comment by Jalle19** [comment]
I'll try ;-)

**Comment by ksooo** [comment]
I trust in you. ;-)
